core/src/main/java/org/infinispan/functional/impl/SimpleReadWriteMapImpl.java

```diff
@@ -0,0 +1,195 @@
+
+   @Override
+   public <T, R> CompletableFuture<R> eval(K key, T argument, BiFunction<T, EntryView.ReadWriteEntryView<K, V>, R> f) {
+      K storageKey = storageKey(key);
+      InternalCacheEntry<K, V> ice = fmap.cache.getDataContainer().peek(storageKey);
+      if (canHandleEntry(ice)) {
+         try {
+            R result = evalInternal(storageKey, null, f);
+            if (result == null) return CompletableFutures.completedNull();
+            return CompletableFuture.completedFuture(result);
+         } catch (Throwable t) {
+            return CompletableFuture.failedFuture(t);
+         }
+      }
+
+      int segment = fmap.keyPartitioner.getSegment(storageKey);
+      return expirationManager.handlePossibleExpiration(ice, segment, true)
+            .thenApply(expired -> {
+               InternalCacheEntry<K, V> entry = !expired ? ice : null;
+               ByRef<R> result = new ByRef<>(null);
+               fmap.cache.getDataContainer().compute(storageKey, (k, old, factory) -> compute(storageKey, null, f, entry, factory, result));
+               return result.get();
+            }).toCompletableFuture();
+   }
+
+   @Override
+   public <T, R> Traversable<R> evalMany(Map<? extends K, ? extends T> arguments, BiFunction<T, EntryView.ReadWriteEntryView<K, V>, R> f) {
+      // Since we're in a simple cache, everything is local!
+      List<R> results = new ArrayList<>(arguments.size());
+      for (Map.Entry<? extends K, ? extends T> me : arguments.entrySet()) {
+         CompletableFuture<R> cf = eval(me.getKey(), me.getValue(), f);
+         assert cf.isDone() : "eval() returned an incomplete CompletableFuture";
+
+         results.add(CompletionStages.join(cf));
+      }
+
+      return Traversables.of(results.stream());
+   }
+
+   @Override
+   public <R> Traversable<R> evalMany(Set<? extends K> keys, Function<EntryView.ReadWriteEntryView<K, V>, R> f) {
+      List<R> results = new ArrayList<>(keys.size());
+      BiFunction<?, EntryView.ReadWriteEntryView<K, V>, R> bf = (ignore, v) -> f.apply(v);
+      for (K key : keys) {
+         CompletableFuture<R> cf = eval(key, null, bf);
+         assert cf.isDone() : "eval() returned an incomplete CompletableFuture";
+
+         results.add(CompletionStages.join(cf));
+      }
+
+      return Traversables.of(results.stream());
+   }
+
+   @Override
+   public <R> Traversable<R> evalAll(Function<EntryView.ReadWriteEntryView<K, V>, R> f) {
+      return evalMany(fmap.cache.keySet(), f);
+   }
+
+   private boolean isStatisticsEnabled() {
+      return !params.containsAll(Param.StatisticsMode.SKIP);
+   }
+
+   private boolean hasListeners() {
+      return !notifier.getListeners().isEmpty();
+   }
+
+   private boolean canHandleEntry(InternalCacheEntry<K, V> ice) {
+      if (ice == null) {
+         return true;
+      }
+
+      if (ice.canExpire()) {
+         if (ice.isExpired(timeService.wallClockTime())) {
+            if (notifier.hasListener(CacheEntryExpired.class)) {
```

Comment:
Expiration events are post only, so this has to go after the entry has been removed.

Reply:
Updated.